### PR TITLE
Add empty cache

### LINKED ISSUE #134332: How do I call empty_cache for the host caching allocator from Python?
### 🐛 Describe the bug

I'm looking for a python binding for the host allocator's `empty_cache` to clear CPU pinned memory. Right now that memory is still held after training finishes, so I'm unable to run any memory intensive post processing. Full context in [this thread](https://github.com/pytorch/pytorch/issues/134051)

cc: @awgu 

### Versions

Collecting environment information...

Vendor ID: GenuineIntel
CPU family: 6
Model: 143
Model name: Intel(R) Xeon(R) Platinum 8480+
Stepping: 8
CPU MHz: 2000.000
CPU max MHz: 3800.0000
CPU min MHz: 800.0000
BogoMIPS: 4000.00
Virtualization: VT-x
L1d cache: 5.3 MiB
L1i cache: 3.5 MiB
L2 cache: 224 MiB
L3 cache: 210 MiB
NUMA node0 CPU(s): 0-55,112-167
NUMA node1 CPU(s): 56-111,168-223
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Vulnerable

pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.16.2
[pip3] onnxruntime==1.18.1
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.3.1+cu121
[pip3] torch-optimizer==0.3.0
[pip3] torchmetrics==1.4.0.post0
[pip3] torchvision==0.18.1+cu121
[pip3] triton==2.3.1
[conda] Could not collect

cc @ptrblck @msaroufim

### PR DESCRIPTION
**Pull request by b-chu**
Fixes #134332

### CODE REVIEW

File: torch/csrc/cuda/Module.cpp [hunk 1]
```diff
@@ -1091,6 +1091,9 @@ static void registerCudaPluggableAllocator(PyObject* module) {
   m.def("_cuda_getAllocator", []() {
     return py::cast(torch::cuda::CUDAPluggableAllocator::getCurrentAllocator());
   });
+  m.def("_cuda_host_empty_cache", []() {
```
Comment (janeyx99) [inline torch/csrc/cuda/Module.cpp:1094]: Should this follow the snake case similar to surrounding functions?

Comment (b-chu) [reply torch/csrc/cuda/Module.cpp:1094]: It'd be great if you took over the PR, I have a testing script in #134332 that reproduces the issue

**Comment by janeyx99** [comment]
I currently don't have the bandwidth to take over the PR, but let's leave the linked issue open so others looking for actionable issues can pick it up.

**Comment by b-chu** [comment]
Closing this since we found a workaround, but would still appreciate a fix

**Pull request closed without merging**